File: app/src/main/java/hu/epam/worktime/mvvmworkdroid/modules/main/model/MainModel.java
package hu.epam.worktime.mvvmworkdroid.modules.main.model;

import android.util.Log;
import hu.epam.worktime.mvvmworkdroid.common.mvvm.Model;
import hu.epam.worktime.mvvmworkdroid.modules.services.WorkServiceApi;
import hu.epam.worktime.mvvmworkdroid.modules.services.models.WorkTime;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import java.util.ArrayList;
import java.util.List;

/**
 *
 *
 * Created by Mihaly_Hunyady on 2016. 12. 02..
 */

public class MainModel implements Model {
    private ModelCallback callback;
    private final WorkServiceApi workServiceApi;
    private List<WorkTime> workTimes = new ArrayList<>();

    public MainModel(WorkServiceApi workServiceApi) {
        this.workServiceApi = workServiceApi;
    }

    public void loadWorkTimes() {
        workServiceApi.workTimes(2).enqueue(new Callback<List<WorkTime>>() {
            @Override
            public void onResponse(Call<List<WorkTime>> call, Response<List<WorkTime>> response) {
                onResult(response);
            }

            @Override
            public void onFailure(Call<List<WorkTime>> call, Throwable t) {
                onError(t);
            }
        });
    }

    private void onResult(Response<List<WorkTime>> workTimes) {
        this.workTimes = workTimes.body();
        onCompleted();
    }

    private void onError(Throwable throwable) {
        Log.d("WorkDroid", "hiba van:" + throwable.getMessage());
    }

    private void onCompleted() {
        if (callback != null) {
            callback.onWorkTimeLoadCompleted();
        }
    }

    public void setCallback(ModelCallback callback) {
        this.callback = callback;
        if (this.callback != null && !workTimes.isEmpty()) {
            this.callback.onWorkTimeLoaded(workTimes);
        }
    }

    public List<WorkTime> getWorkTimes() {
        return workTimes;
    }

    public interface ModelCallback {
        void onWorkTimeLoaded(List<WorkTime> workTimes);

        void onWorkTimeLoadError();

        void onWorkTimeLoadCompleted();
    }
}

File: app/src/main/java/hu/epam/worktime/mvvmworkdroid/modules/main/main/model/MainModel.kt
package hu.epam.worktime.mvvmworkdroid.modules.main.main.model

import hu.epam.worktime.mvvmworkdroid.modules.dal.WorkItemDao
import hu.epam.worktime.mvvmworkdroid.modules.services.WorkServiceApi
import hu.epam.worktime.mvvmworkdroid.modules.services.models.WorkDay
import hu.epam.worktime.mvvmworkdroid.modules.services.models.WorkTime
import hu.epam.worktime.mvvmworkdroid.modules.services.models.WorkingStatistics
import hu.epam.worktime.mvvmworkdroid.modules.services.worker.CalculatorService
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by Mihaly_Hunyady on 2016. 12. 12..
 */

class MainModel(private val workServiceApi: WorkServiceApi, private val calculatorService: CalculatorService, private val workItemDao: WorkItemDao) {
    var callback: MainModel.ModelCallback? = null
    var workingStatistics: WorkingStatistics? = null
    private lateinit var workTimes: List<WorkTime>

    fun loadWorkTimes() {
        workServiceApi.workTimes(2)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ onResult(it) }, { onError(it) }) { onWorkTimesCompleted() }
    }

    private fun loadWorkDays() {
        workServiceApi.workDays()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ onWorkDayResult(it) }, { onError(it) }) { onCompleted() }
    }

    private fun onWorkDayResult(workDays: List<WorkDay>?) {
        if (workDays != null && workDays.isNotEmpty()) {
            calculatorService.workDays = workDays
            this.workingStatistics = calculatorService.calculateStuffs(workTimes)
            workingStatistics?.let {
                workItemDao.saveWorkingStatistics(workingStatistics!!)
            }
        }

        callback?.run { onWorkTimeLoaded(workItemDao.find()) }

    }

    private fun onResult(workTimes: List<WorkTime>) {
        this.workTimes = workTimes
        loadWorkDays()
    }

    private fun onError(throwable: Throwable) {
        workingStatistics = workItemDao.find()
        callback?.run {
            onWorkTimeLoadError(throwable)
        }
    }

    private fun onWorkTimesCompleted() {
        loadWorkDays()
    }

    private fun onCompleted() {
        callback?.run {
            onWorkTimeLoadCompleted()
        }
    }

    interface ModelCallback {
        fun onWorkTimeLoaded(workTimes: WorkingStatistics)

        fun onWorkTimeLoadError(throwable: Throwable)

        fun onWorkTimeLoadCompleted()
    }
}
